packages/checkout-storefront/src/hooks/useLocale.ts

```diff
@@ -0,0 +1,53 @@
+import { CountryCode } from "@/checkout-storefront/graphql";
+import { UrlChangeHandlerArgs, useUrlChange } from "@/checkout-storefront/hooks/useUrlChange";
+import { DEFAULT_LOCALE, Locale } from "@/checkout-storefront/lib/regions";
+import { getParsedLocaleData, getQueryParams } from "@/checkout-storefront/lib/utils";
+import { useMemo, useState } from "react";
+
+import EN_US from "../../content/compiled-locales/en-US.json";
+import PL_PL from "../../content/compiled-locales/pl-PL.json";
+import Minion from "../../content/compiled-locales/minion.json";
+
+const localeToMessages: Record<Locale, any> = {
+  "en-US": EN_US,
+  "pl-PL": PL_PL,
+  minion: Minion,
+};
+
+interface UseLocale {
+  locale: Locale;
+  countryCode: CountryCode;
+  messages: typeof localeToMessages[keyof typeof localeToMessages];
+}
+
+export const useLocale = (): UseLocale => {
+  const { locale, countryCode } = getParsedLocaleData(getQueryParams().locale);
+
+  const [currentLocale, setCurrentLocale] = useState<Locale>(locale);
```

Comment:
I think we do 😛 The thing that causes the rerender is setState we do in handler for listener on window event `popstate`. If we only set state for `countryCode` then it works 99% of times, but then we get locales like `en-CA` and `fr-CA` (both valid languages for Canada 🇨🇦) -> gthen the languageCode doesn't update because setState doesn't happen underneath cause the country code stays the same

Reply:
It sounds like a bug in `useUrlChange` 🤔

Reply:
useUrlChange fires when there was a change in the url (either countryCode or languageCode). But if in `useLocale` we don't change the state based on this being called then...well we don't get a rerender. Seems like react rather than a bug 😅

Reply:
Anyway, let's leave it as is for now 👍 and I'll think about it